x/bridge/observer/osmosis/osmosis.go

```diff
@@ -0,0 +1,208 @@
+	bridgetypes "github.com/osmosis-labs/osmosis/v24/x/bridge/types"
+)
+
+var (
+	ModuleName    = "osmosis-chain"
+	OsmoGasLimit  = uint64(200000)
+	OsmoFeeAmount = sdktypes.NewIntFromUint64(1000)
+	OsmoFeeDenom  = "uosmo"
+)
+
+type ChainClient struct {
+	logger             log.Logger
+	osmoClient         *Client
+	cometRpc           *rpchttp.HTTP
+	stopChan           chan struct{}
+	outboundChan       chan observer.Transfer
+	lastObservedHeight atomic.Uint64
+}
+
+// NewChainClient returns new instance of `Osmosis`
+func NewChainClient(
+	logger log.Logger,
+	osmoClient *Client,
+	cometRpc *rpchttp.HTTP,
+) *ChainClient {
+	return &ChainClient{
+		logger:             logger.With("module", ModuleName),
+		osmoClient:         osmoClient,
+		cometRpc:           cometRpc,
+		stopChan:           make(chan struct{}),
+		outboundChan:       make(chan observer.Transfer),
+		lastObservedHeight: atomic.Uint64{},
+	}
+}
+
+// Start subscribes to the `NewBlock` events and starts listening to `EventOutboundTransfer` events
+func (o *ChainClient) Start(ctx context.Context) error {
+	err := o.cometRpc.Start()
+	if err != nil {
+		return errorsmod.Wrapf(ErrRpcClient, err.Error())
+	}
+
+	txs, err := o.cometRpc.Subscribe(ctx, ModuleName, comettypes.EventQueryNewBlock.String())
+	if err != nil {
+		return errorsmod.Wrapf(ErrRpcClient, err.Error())
+	}
+
+	go o.observeEvents(ctx, txs)
+
+	return nil
+}
+
+// Stop stops listening to events and closes Osmosis client
+func (o *ChainClient) Stop(ctx context.Context) error {
+	close(o.stopChan)
+	o.osmoClient.Close()
+	if err := o.cometRpc.UnsubscribeAll(ctx, ModuleName); err != nil {
+		return errorsmod.Wrapf(err, "Failed to unsubscribe from RPC client")
+	}
+	return o.cometRpc.Stop()
+}
+
+// ListenOutboundTransfer returns receive-only channel with `OutboundTransfer` items
+func (o *ChainClient) ListenOutboundTransfer() <-chan observer.Transfer {
+	return o.outboundChan
+}
+
+// SignalInboundTransfer sends `InboundTransfer` to Osmosis
+func (o *ChainClient) SignalInboundTransfer(ctx context.Context, in observer.Transfer) error {
+	msg := bridgetypes.NewMsgInboundTransfer(
+		in.Id,
+		in.Sender,
+		in.To,
+		bridgetypes.AssetID{
+			SourceChain: string(in.SrcChain),
+			Denom:       in.Asset,
+		},
+		math.Int(in.Amount),
+	)
+	fees := sdktypes.NewCoins(sdktypes.NewCoin(OsmoFeeDenom, OsmoFeeAmount))
+	bytes, err := o.osmoClient.SignTx(ctx, msg, fees, OsmoGasLimit)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to sign tx for inbound transfer %s", in.Id)
+	}
+	_, err = o.osmoClient.BroadcastTx(ctx, bytes)
+	if err != nil {
+		return errorsmod.Wrapf(
+			err,
+			"Failed to broadcast tx to Osmosis for inbound transfer %s",
+			in.Id,
+		)
+	}
+	return nil
+}
+
+func (o *ChainClient) observeEvents(
+	ctx context.Context,
+	txs <-chan coretypes.ResultEvent,
+) {
+	defer close(o.outboundChan)
+
+	eventToObserve := cosmosproto.MessageName(&bridgetypes.EventOutboundTransfer{})
```

Comment:
I think this needs to change from an event observer to a transaction type observer. There have been historical examples across multiple cosmos chains (including Osmosis) where we accidentally duplicate events, which would be catastrophic in this use case. Happy to discuss offline way forward if not clear